gateway/src/shard/emitter.rs

```diff
@@ -0,0 +1,184 @@
+use super::json;
+use crate::{Event, EventTypeFlags};
+use std::{
+    convert::TryFrom,
+    error::Error,
+    fmt::{Display, Formatter, Result as FmtResult},
+};
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use twilight_model::gateway::event::shard::Payload;
+
+#[derive(Debug)]
+pub struct EmitJsonError {
+    kind: EmitJsonErrorType,
+    source: Option<Box<dyn Error + Send + Sync>>,
+}
+
+impl EmitJsonError {
+    pub fn into_parts(self) -> (EmitJsonErrorType, Option<Box<dyn Error + Send + Sync>>) {
+        (self.kind, self.source)
+    }
+}
+
+impl Display for EmitJsonError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match &self.kind {
+            EmitJsonErrorType::EventTypeUnknown { event_type, op } => f.write_fmt(format_args!(
+                "provided event type ({:?})/op ({}) pair is unknown",
+                event_type, op,
+            )),
+            EmitJsonErrorType::Parsing => f.write_str("parsing a gateway event failed"),
+        }
+    }
+}
+
+impl Error for EmitJsonError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
+    }
+}
+
+/// Type of [`EmitJsonError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum EmitJsonErrorType {
+    /// Provided event type and/or opcode combination doesn't match a known
+    /// event type flag.
+    EventTypeUnknown {
+        /// Received dispatch event type.
+        event_type: Option<String>,
+        /// Received opcode.
+        op: u8,
+    },
+    /// Parsing a a gateway event failed.
+    Parsing,
+}
+
+/// Emitter over a listener with some useful things on top to abstract common
+/// operations.
+#[derive(Clone, Debug)]
+pub struct Emitter {
+    event_types: EventTypeFlags,
+    tx: UnboundedSender<Event>,
+}
+
+impl Emitter {
```

Comment:
`Config` has an `event_types` function. Would that have any use here as well to see the internal filter state?